### PR TITLE
feat: cleaner wrapper

### PR DESCRIPTION
**Pull request by SimonDuToit**
Jumanji wrapper and scenarios for Cleaner environment.

### CODE REVIEW

File: mava/configs/env/cleaner.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# Defines the metric that will be used to evaluate the performance of the agent.
+# This metric is returned at the end of an experiment and can be used for hyperparameter tuning.
+eval_metric: episode_return
+
+# Whether the environment observations encode implicit agent IDs. If True, the AgentID wrapper is not used.
+# This should not be changed.
+implicit_agent_id: True
+
+kwargs:
+  time_limit: 100
```
Comment (sash-a) [inline mava/configs/env/cleaner.yaml:18]: I worry about this being an argument of the env. Although as long as the reward encorages agents to finish as soon as possible time limit shouldn't matter right?

File: mava/wrappers/jumanji.py [hunk 2]
```diff
@@ -265,3 +267,104 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+    """Multi-agent wrapper for the Cleaner environment."""
+
+    def __init__(self, env: Cleaner, add_global_state: bool = False):
+        super().__init__(env, add_global_state)
+        self._env: Cleaner
+
+    def modify_timestep(self, timestep: TimeStep) -> TimeStep[Observation]:
+        """Modify the timestep for the Cleaner environment."""
+
+        def create_agents_view(grid: chex.Array, agents_locations: chex.Array) -> chex.Array:
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:282]: These comments are amazing, makes it so easy to follow! Awesome work here :fire:

File: mava/wrappers/jumanji.py [hunk 3]
```diff
@@ -265,3 +267,104 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+            # C: Number of grid columns
+            # grid: (R, C)
+
+            # (N, 2)
+            agents_locations = agents_locations
+
+            # Get dirty / wall tiles from first agent's obs and tile.
+            # (A, R, C)
+            dirty_channel = jnp.tile(jnp.where(grid == DIRTY, 1, 0), (num_agents, 1, 1))
+            wall_channel = jnp.tile(jnp.where(grid == WALL, 1, 0), (num_agents, 1, 1))
```
Comment (sash-a) [inline mava/wrappers/jumanji.py]: No need for the `where`
```suggestion
            dirty_channel = jnp.tile(grid == DIRTY, (num_agents, 1, 1))
            wall_channel = jnp.tile(grid == WALL, (num_agents, 1, 1))
```

File: mava/wrappers/jumanji.py [hunk 4]
```diff
@@ -265,3 +267,104 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+
+            # (2, N)
+            xs, ys = agents_locations[:, 0], agents_locations[:, 1]
+
+            # Mask each agent's position so an agent can idenfity itself.
+            # Sum the masked grids together for global agent information.
+            # (A, R, C)
+            pos_per_agent = jnp.repeat(jnp.zeros_like(grid)[None, :, :], num_agents, axis=0)
```
Comment (sash-a) [inline mava/wrappers/jumanji.py]: `jnp.newaxis` is a bit clearer than `None`
```suggestion
            pos_per_agent = jnp.repeat(jnp.zeros_like(grid)[jnp.newaxis, :, :], num_agents, axis=0)
```

File: mava/wrappers/jumanji.py [hunk 5]
```diff
@@ -265,3 +267,104 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+            pos_per_agent = pos_per_agent.at[jnp.arange(num_agents), xs, ys].set(1)
+            agents_channel = jnp.tile(jnp.sum(pos_per_agent, axis=0), (num_agents, 1, 1))
```
Comment (sash-a) [inline mava/wrappers/jumanji.py:309]: Is this so every agent knows where every other agent is? Do we want this because then it's not partially observable, although the env isn't PO to begin with :thinking:

**Review by sash-a — CHANGES_REQUESTED** [review]
Really awesome work! Just some very minor questions and clean up

File: mava/configs/env/cleaner.yaml (hunk 1 above)
Comment (SimonDuToit) [reply mava/configs/env/cleaner.yaml:18]: I agree the time limit should be a scenario argument. But Ill make a separate issue for this since it involves changing all the other envs too.

File: mava/wrappers/jumanji.py (hunk 2 above)
Comment (SimonDuToit) [reply mava/wrappers/jumanji.py:282]: Thanks. Ill add these to connector eventually too, theres a few small things there I want to change.

File: mava/wrappers/jumanji.py (hunk 5 above)
Comment (SimonDuToit) [reply mava/wrappers/jumanji.py:309]: Yes. And yes I dont want to lose information from the original env. I think it would be very hard to learn if an agent cant see the other agents. So each agent sees a summary of all the agent positions and its own location.

File: mava/configs/env/scenario/clean-20x20x20a.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# The config of the 20x20x40a scenario
+name: Cleaner-v0
+task_name: clean-20x20x40a
+
+task_config:
+  num_rows: 20
+  num_cols: 20
+  num_agents: 40
```
Comment (sash-a) [inline mava/configs/env/scenario/clean-20x20x20a.yaml]: Think this is a typo
```suggestion
  num_agents: 20
```

File: mava/configs/env/cleaner.yaml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# ---Environment Configs---
+defaults:
+  - _self_
+  - scenario: clean-5x5x5a # [clean-10x10x10a, clean-15x15x15a, clean-20x20x20a, clean-30x30x30a]
```
Comment (sash-a) [inline mava/configs/env/cleaner.yaml]: also gotta add in this option
```suggestion
  - scenario: clean-5x5x5a # [clean-5x5x5a, clean-10x10x10a, clean-15x15x15a, clean-20x20x20a, clean-30x30x30a]
```

File: mava/wrappers/jumanji.py [hunk 8]
```diff
@@ -265,3 +267,103 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+            # grid: (R, C)
+            # agents_locations: (N, 2)
+
+            # Get dirty / wall tiles from first agent's obs and tile.
+            # (A, R, C)
+            dirty_channel = jnp.tile(grid == DIRTY, (num_agents, 1, 1))
+            wall_channel = jnp.tile(grid == WALL, (num_agents, 1, 1))
+
+            # Get each agent's position.
+            # (2, N)
```
Comment (sash-a) [inline mava/wrappers/jumanji.py]: Seems like you're using both A and N here to represent num agents, we usually use A
```suggestion
            # A: Number of agents
            # R: Number of grid rows
            # C: Number of grid columns
            # grid: (R, C)
            # agents_locations: (A, 2)

            # Get dirty / wall tiles from first agent's obs and tile.
            # (A, R, C)
            dirty_channel = jnp.tile(grid == DIRTY, (num_agents, 1, 1))
            wall_channel = jnp.tile(grid == WALL, (num_agents, 1, 1))

            # Get each agent's position.
            # (2, A)
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Looks great couple minor things, but otherwise it's ready to go :smile:

File: mava/wrappers/jumanji.py (hunk 8 above)
Comment (SimonDuToit) [reply mava/wrappers/jumanji.py]: Good catch, thanks

File: mava/wrappers/jumanji.py [hunk 9]
```diff
@@ -265,3 +267,103 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+        reward = jnp.repeat(timestep.reward, self.num_agents)
+        discount = jnp.repeat(timestep.discount, self.num_agents)
+
+        return timestep.replace(
+            observation=Observation(**obs_data), reward=reward, discount=discount
+        )
+
+    def get_global_state(self, obs: Observation) -> chex.Array:
+        """Constructs the global state from the global information
+        in the agent observations (dirty tiles, wall tiles and agent positions)."""
```
Comment (sash-a) [inline mava/wrappers/jumanji.py]: nit
```suggestion
        """Constructs the global state from the global information
        in the agent observations (dirty tiles, wall tiles and agent positions).
        """
```

File: mava/wrappers/jumanji.py [hunk 10]
```diff
@@ -265,3 +267,103 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+    def __init__(self, env: Cleaner, add_global_state: bool = False):
+        super().__init__(env, add_global_state)
+        self._env: Cleaner
+
+    def modify_timestep(self, timestep: TimeStep) -> TimeStep[Observation]:
+        """Modify the timestep for the Cleaner environment."""
+
+        def create_agents_view(grid: chex.Array, agents_locations: chex.Array) -> chex.Array:
+            """Create separate channels for dirty cells, wall cells and agent positions.
+            Also add a channel that marks an agent's own position."""
```
Comment (sash-a) [inline mava/wrappers/jumanji.py]:
```suggestion
            """Create separate channels for dirty cells, wall cells and agent positions.
            Also add a channel that marks an agent's own position.
            """
```

File: mava/wrappers/jumanji.py [hunk 11]
```diff
@@ -265,3 +267,103 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+            # A: Number of agents
+            # R: Number of grid rows
+            # C: Number of grid columns
+            # grid: (R, C)
+            # agents_locations: (A, 2)
+
+            # Get dirty / wall tiles from first agent's obs and tile in agents dimension.
+            # (A, R, C)
+            dirty_channel = jnp.tile(grid == DIRTY, (num_agents, 1, 1))
+            wall_channel = jnp.tile(grid == WALL, (num_agents, 1, 1))
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py]: It is a bit easier to read when the a shape is on the same line.
```suggestion
            dirty_channel = jnp.tile(grid == DIRTY, (num_agents, 1, 1))  # (A, R, C)
            wall_channel = jnp.tile(grid == WALL, (num_agents, 1, 1))  # (A, R, C)
```

File: mava/wrappers/jumanji.py [hunk 12]
```diff
@@ -265,3 +267,103 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+
+            # Get each agent's position.
+            # (2, A)
+            xs, ys = agents_locations[:, 0], agents_locations[:, 1]
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py]: Similarly for the shapes here please. Also since these are two arrays please have a shape per array.

File: mava/wrappers/jumanji.py [hunk 13]
```diff
@@ -265,3 +267,103 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+
+            # Mask each agent's position so an agent can idenfity itself.
+            # Sum the masked grids together for global agent information.
+            # (A, R, C)
+            pos_per_agent = jnp.repeat(jnp.zeros_like(grid)[jnp.newaxis, :, :], num_agents, axis=0)
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py:306]: Similarly for the shapes here and elsewhere please :pray:

File: mava/wrappers/jumanji.py [hunk 14]
```diff
@@ -265,3 +267,103 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+        discount = jnp.repeat(timestep.discount, self.num_agents)
+
+        return timestep.replace(
+            observation=Observation(**obs_data), reward=reward, discount=discount
+        )
+
+    def get_global_state(self, obs: Observation) -> chex.Array:
+        """Constructs the global state from the global information
+        in the agent observations (dirty tiles, wall tiles and agent positions)."""
+        return obs.agents_view[..., :3]
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py]: Shape here please?

**Review by RuanJohn — CHANGES_REQUESTED** [review]
Thank you @SimonDuToit! This is looking pretty much ready to go! Just some comments around how we show shapes.

File: mava/wrappers/jumanji.py (hunk 11 above)
Comment (SimonDuToit) [reply mava/wrappers/jumanji.py]: On some of these lines, putting the comment on the same line causes the precommit to to turn it into 3 lines. Is this still preferable? Example:
image

Comment (RuanJohn) [reply mava/wrappers/jumanji.py]: Yep, this is fine.

File: mava/wrappers/jumanji.py [hunk 15]
```diff
@@ -265,3 +267,105 @@ def observation_spec(self) -> specs.Spec[Union[Observation, ObservationGlobalSta
...
+            # grid: (R, C)
+            # agents_locations: (A, 2)
+
+            # Get dirty / wall tiles from first agent's obs and tile in agents dimension.
+
+            dirty_channel = jnp.tile(grid == DIRTY, (num_agents, 1, 1))  # (A, R, C)
+            wall_channel = jnp.tile(grid == WALL, (num_agents, 1, 1))  # (A, R, C)
+
+            # Get each agent's position.
+            xs, ys = agents_locations[:, 0], agents_locations[:, 1]  # (A), (A)
```
Comment (RuanJohn) [inline mava/wrappers/jumanji.py]:
```suggestion
            xs, ys = agents_locations[:, 0], agents_locations[:, 1]  # (A,), (A,)
```

**Review by RuanJohn — DISMISSED** [review]
Can confirm, runs fine for me. 
Thank you for this @SimonDuToit :fire:

File: mava/wrappers/jumanji.py [hunk 16]
```diff
@@ -292,25 +292,27 @@ def create_agents_view(grid: chex.Array, agents_locations: chex.Array) -> chex.A
...
 
             # Mask each agent's position so an agent can idenfity itself.
             # Sum the masked grids together for global agent information.
-            # (A, R, C)
-            pos_per_agent = jnp.repeat(jnp.zeros_like(grid)[jnp.newaxis, :, :], num_agents, axis=0)
-            pos_per_agent = pos_per_agent.at[jnp.arange(num_agents), xs, ys].set(1)
-            agents_channel = jnp.tile(jnp.sum(pos_per_agent, axis=0), (num_agents, 1, 1))
+            pos_per_agent = jnp.repeat(
+                jnp.zeros_like(grid)[jnp.newaxis, :, :], num_agents, axis=0
+            )  # (A, R, C)
```
Comment (sash-a) [inline mava/wrappers/jumanji.py]: Sorry @RuanJohn and I disagree on this :joy:
Resolved to keep shape comments above if it doesn't cause the line to have to reformat into multiple lines
```suggestion
            # (A, R, C)
            pos_per_agent = jnp.repeat(jnp.zeros_like(grid)[jnp.newaxis, :, :], num_agents, axis=0)
```

File: mava/wrappers/jumanji.py [hunk 17]
```diff
@@ -292,25 +292,27 @@ def create_agents_view(grid: chex.Array, agents_locations: chex.Array) -> chex.A
...
+            pos_per_agent = pos_per_agent.at[jnp.arange(num_agents), xs, ys].set(1)  # (A, R, C)
+            agents_channel = jnp.tile(
+                jnp.sum(pos_per_agent, axis=0), (num_agents, 1, 1)
+            )  # (A, R, C)
```
Comment (sash-a) [inline mava/wrappers/jumanji.py]:
```suggestion
            # (A, R, C)
            agents_channel = jnp.tile(jnp.sum(pos_per_agent, axis=0), (num_agents, 1, 1))
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Just these comments I think all can be easily accepted

**Review by sash-a — APPROVED** [review]

**Review by sash-a — CHANGES_REQUESTED** [review]
As discussed offline can we add in a win rate in `extras["won_episode"]`. See the gigastep and smax wrappers for how they do it

**Review by sash-a — APPROVED** [review]
Thanks for this!

**Review by OmaymaMahjoub — APPROVED** [review]
LGTM!
